src/netbsd.rs

```diff
@@ -25,27 +41,33 @@ fn kern_arnd(buf: &mut [MaybeUninit<u8>]) -> libc::ssize_t {
 type GetRandomFn = unsafe extern "C" fn(*mut u8, libc::size_t, libc::c_uint) -> libc::ssize_t;
 
 // getrandom(2) was introduced in NetBSD 10.0
-static GETRANDOM: LazyPtr = LazyPtr::new();
+static GETRANDOM: AtomicPtr<c_void> = AtomicPtr::new(ptr::null_mut());
 
-fn dlsym_getrandom() -> *mut c_void {
+#[cold]
+fn init() -> *mut c_void {
     static NAME: &[u8] = b"getrandom\0";
     let name_ptr = NAME.as_ptr().cast::<libc::c_char>();
-    unsafe { libc::dlsym(libc::RTLD_DEFAULT, name_ptr) }
+    let ptr = match unsafe { libc::dlsym(libc::RTLD_DEFAULT, name_ptr) } {
+        ptr if !ptr.is_null() => ptr,
+        _ => polyfill_using_kern_arand as *mut libc::c_void,
+    };
+    GETRANDOM.store(ptr, Ordering::Release);
+    ptr
 }
 
 pub fn getrandom_inner(dest: &mut [MaybeUninit<u8>]) -> Result<(), Error> {
-    let fptr = GETRANDOM.unsync_init(dlsym_getrandom);
-    if !fptr.is_null() {
-        let func: GetRandomFn = unsafe { core::mem::transmute(fptr) };
-        return sys_fill_exact(dest, |buf| unsafe {
-            func(buf.as_mut_ptr().cast::<u8>(), buf.len(), 0)
-        });
-    }
-
-    // NetBSD will only return up to 256 bytes at a time, and
-    // older NetBSD kernels will fail on longer buffers.
-    for chunk in dest.chunks_mut(256) {
-        sys_fill_exact(chunk, kern_arnd)?
-    }
-    Ok(())
+    // Despite being only a single atomic variable, we still cannot always use
+    // Ordering::Relaxed, as we need to make sure a successful call to `init`
+    // is "ordered before" any data read through the returned pointer (which
+    // occurs when the function is called). Our implementation mirrors that of
+    // the one in libstd, meaning that the use of non-Relaxed operations is
+    // probably unnecessary.
+    let fptr = match GETRANDOM.load(Ordering::Acquire) {
+        fptr if !fptr.is_null() => fptr,
+        _ => init(),
+    };
+    let fptr = unsafe { transmute::<*mut c_void, GetRandomFn>(fptr) };
+    sys_fill_exact(dest, |buf| unsafe {
+        fptr(buf.as_mut_ptr().cast::<u8>(), buf.len(), 0)
```

Comment:
`cast` is not necessary on this line, since `MaybeUninit::as_mut_ptr` returns `*mut T`.

Reply:
This is `[MaybeUninit<[u8]>::as_mut_ptr()`, i.e. `slice::as_mut_ptr()`, not `MaybeUninit::as_mut_ptr`.

Reply:
Ah, true.